pkg/teleworker/job.go

```diff
@@ -0,0 +1,186 @@
+	}
+
+	return formatted
+}
+
+func (l *Limits) ToFlags() []string {
+	flags := []string{}
+	if l.MemoryMB > 0 {
+		flags = append(flags, fmt.Sprintf("-memorymb=%d", l.MemoryMB))
+	}
+
+	if l.CpuWeight > 0 {
+		flags = append(flags, fmt.Sprintf("-cpuweight=%d", l.CpuWeight))
+	}
+
+	if l.IOWeight > 0 {
+		flags = append(flags, fmt.Sprintf("-ioweight=%d", l.IOWeight))
+	}
+	return flags
+}
+
+type JobState struct {
+	Status   api.JobStatus
+	ExitCode int
+	ExitErr  error
+	ExitedAt time.Time
+	Limits   *Limits
+}
+
+// StatusName returns user friendly name for JobStatus
+// See api.JobStatus_name for code to string mapping
+func (js *JobState) StatusName() string {
+	name, ok := api.JobStatus_name[int32(js.Status)]
+	if !ok {
+		// Fallback to the default UNKNOWN status,
+		// but should never happen in reality
+		return api.JobStatus_name[0]
+	}
+	return name
+}
+
+type Job struct {
+	ID          uuid.UUID
+	UserCommand string
+	UserArgs    []string
+
+	mu        sync.RWMutex
+	cmd       *exec.Cmd
+	outLogger *ls.LogStreamer
+	errLogger *ls.LogStreamer
+
+	user  string
+	state *JobState
+
+	done chan struct{}
+}
+
+type Option func(*Job)
+
+func NewJob(command string, args []string, options ...Option) (*Job, error) {
+	uuid, err := uuid.NewV4()
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error generating uuid: %w", err)
+	}
+
+	j := &Job{
+		ID:          uuid,
+		UserCommand: command,
+		UserArgs:    args,
+
+		state: &JobState{
+			Limits: &Limits{},
+			Status: api.JobStatus_STARTING,
+		},
+
+		done: make(chan struct{}),
+	}
+
+	// It is easy to imagine that in some cases it is required to use
+	// the library without, for example, resource limits, so some of the
+	// parameters are optional and can be omitted when configuring the server
+	for _, opt := range options {
+		opt(j)
+	}
+
+	cmd := j.selfWrapCommand()
+
+	outReader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("error setting up stdout logger: %w", err)
+	}
+	errReader, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, fmt.Errorf("error setting up stderr logger: %w", err)
+	}
+
+	j.cmd = cmd
+	j.outLogger = ls.NewLogStreamer(outReader)
+	j.errLogger = ls.NewLogStreamer(errReader)
+
+	return j, nil
+}
+
+// selfWrapCommand wraps the user command in the self call.
+// Whenever user wants to start some task we launch another server process and
+// distinguish it from the main flow using flags. Then we can use that process ID
+// in order to limit the resources of user's task before we actually start it.
+func (j *Job) selfWrapCommand() *exec.Cmd {
+	const selfExe = "/proc/self/exe"
+
+	jobID := fmt.Sprintf("-jobid=%s", j.ID.String())
+	userCommand := fmt.Sprintf("-command=%s", j.UserCommand)
+	limitFlags := j.state.Limits.ToFlags()
+
+	callArgs := append(limitFlags, jobID)
+	callArgs = append(callArgs, userCommand)
+	callArgs = append(callArgs, j.UserArgs...)
+
+	return exec.Command(selfExe, callArgs...)
+}
```

Comment:
How would this work in case of the library? I.e. if I want to import this library in my own program.

Reply:
It refers to the question on cgroups - as we want to manage "put PID to the group before the user's command is actually started", my idea was to make the server responsible for managing all the input the library provides as flags. Giving it a second thought, it would make more sense to also encapsulate all that logic into the library and just make the server being responsible only for using this `tw.UnwrapInternalCall` method at the beginning of the server's `main`. I'll add this method to the package then if it makes sense